presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java

```diff
@@ -0,0 +1,222 @@
+    private final long maxFilterSizeInBytes;
+
+    private final List<Integer> channels;
+    private final List<Type> buildTypes;
+
+    // May be dropped if the predicate becomes too large.
+    private BlockBuilder[] blockBuilders;
+    private TypedSet[] valueSets;
+
+    private int filterPositionsCount;
+    private long filterSizeInBytes;
+
+    public DynamicFilterCollectOperator(OperatorContext context, Consumer<TupleDomain<Integer>> dynamicPredicateConsumer, List<Integer> channels, List<Type> buildTypes, PlanNodeId planNodeId)
+    {
+        this.context = requireNonNull(context, "context is null");
+        this.maxFilterPositionsCount = getDynamicFilteringMaxPerDriverValueCount(context.getSession());
+        this.maxFilterSizeInBytes = getDynamicFilteringMaxPerDriverByteSize(context.getSession()).toBytes();
+
+        this.dynamicPredicateConsumer = requireNonNull(dynamicPredicateConsumer, "dynamicPredicateConsumer is null");
+        this.channels = requireNonNull(channels, "channels is null");
+        this.buildTypes = requireNonNull(buildTypes, "buildTypes is null");
+
+        final int expectedSize = 8;
+        this.blockBuilders = new BlockBuilder[channels.size()];
+        this.valueSets = new TypedSet[channels.size()];
+        for (int channelIndex = 0; channelIndex < channels.size(); ++channelIndex) {
+            Type type = buildTypes.get(channelIndex);
+            this.blockBuilders[channelIndex] = type.createBlockBuilder(null, expectedSize);
+            this.valueSets[channelIndex] = new TypedSet(type, Optional.empty(), blockBuilders[channelIndex], expectedSize,
+                                                        String.format("DynamicFilterCollectOperator_%s_%d", planNodeId, channelIndex));
+        }
+        this.filterPositionsCount = 0;  // non-null values count
+        this.filterSizeInBytes = 0;  // non-null values size in bytes
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return context;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return current == null && !finished;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        verify(!finished, "DynamicFilterCollectOperator: addInput() may not be called after finish()");
+        current = page;
+        if (valueSets == null) {
+            return;  // the predicate became too large.
+        }
+
+        // Collect only the columns which are relevant for the JOIN.
+        for (int channelIndex = 0; channelIndex < channels.size(); ++channelIndex) {
+            Block block = page.getBlock(channels.get(channelIndex));
+            for (int position = 0; position < block.getPositionCount(); ++position) {
+                if (!valueSets[channelIndex].contains(block, position)) {
+                    valueSets[channelIndex].add(block, position);
+                    filterSizeInBytes += block.getEstimatedDataSizeForStats(position);
```

Comment:
You can now use `io.prestosql.operator.aggregation.TypedSet#getRetainedSizeInBytes` after processing whole page. It should be considerable quicker and more accurate

Reply:
Good catch, thanks!
https://github.com/prestosql/presto/blob/8d8f3e66d252d2b1275e639460d966d53f65977a/presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java#L162-L182
WDYT?